Incorrect error code when using a trait object without `dyn`

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.
-->

Hi all, thank you for the great work on rust!

I believe the compiler might be suggesting an incorrect error code when omitting `dyn` with trait objects. An example is the following snippet ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=a0d8ddac00f1103c28563566f8bbe175)):

```rust
trait MyTrait {
}

impl dyn MyTrait {
    fn get_num() -> i32 {
        7
    }
}

fn main() {
    println!("The number is: {}", MyTrait::get_num());
}
```

The error code [E0783](https://doc.rust-lang.org/error-index.html#E0783) in the error message seems to point to an error regarding the range pattern. I believe the correct error may instead be [E0782](https://doc.rust-lang.org/error-index.html#E0782).
```
   Compiling playground v0.0.1 (/playground)
error[E0783]: trait objects without an explicit `dyn` are deprecated
  --> src/main.rs:13:35
   |
13 |     println!("The number is: {}", MyTrait::get_num());
   |                                   ^^^^^^^ help: use `dyn`: `<dyn MyTrait>`

For more information about this error, try `rustc --explain E0783`.
error: could not compile `playground` due to previous error
```

### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
rustc --version --verbose
rustc 1.58.0-nightly (8b09ba6a5 2021-11-09)
binary: rustc
commit-hash: 8b09ba6a5d5c644fe0f1c27c7f9c80b334241707
commit-date: 2021-11-09
host: x86_64-unknown-linux-gnu
release: 1.58.0-nightly
LLVM version: 13.0.0

```

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary>Backtrace</summary>
<p>

```
<backtrace>
```

</p>
</details>

--- a/compiler/rustc_typeck/src/check/fn_ctxt/_impl.rs
+++ b/compiler/rustc_typeck/src/check/fn_ctxt/_impl.rs
@@ -952,7 +952,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                     let mut err = rustc_errors::struct_span_err!(
                         self.sess(),
                         self_ty.span,
-                        E0783,
+                        E0782,
                         "{}",
                         msg,
                     );

--- a/src/test/ui/editions/dyn-trait-sugg-2021.stderr
+++ b/src/test/ui/editions/dyn-trait-sugg-2021.stderr
@@ -1,4 +1,4 @@
-error[E0783]: trait objects without an explicit `dyn` are deprecated
+error[E0782]: trait objects without an explicit `dyn` are deprecated
   --> $DIR/dyn-trait-sugg-2021.rs:10:5
    |
 LL |     Foo::hi(123);
@@ -6,4 +6,4 @@ LL |     Foo::hi(123);
 
 error: aborting due to previous error
 
-For more information about this error, try `rustc --explain E0783`.
+For more information about this error, try `rustc --explain E0782`.

